src/native-shim.js

```diff
@@ -10,23 +10,161 @@
+  const nativeDefine = window.customElements.define;
+  const nativeGet = window.customElements.get;
+
+  /**
+   * Map of user-provided constructors to tag names.
+   *
+   * @type {Map<Function, string>}
+   */
+  const tagnameByConstructor = new Map();
+
+  /**
+   * Map of tag anmes to user-provided constructors.
+   *
+   * @type {Map<string, Function>}
+   */
+  const constructorByTagname = new Map();
+
+
+  /**
+   * Whether the constructors are being called by a browser process, ie parsing
+   * or createElement.
+   */
+  let browserConstruction = false;
+
+  /**
+   * Whether the constructors are being called by a user-space process, ie
+   * calling an element constructor.
+   */
+  let userConstruction = false;
+
   window.HTMLElement = function() {
-    // prefer new.target for elements that call super() constructors or
-    // Reflect.construct directly
-    let newTarget = new.target || this.constructor;
-    return Reflect.construct(origHTMLElement, [], newTarget);
-  }
-  HTMLElement.prototype = Object.create(origHTMLElement.prototype, {
-    constructor: {value: HTMLElement, configurable: true, writable: true},
-  });
+    if (!browserConstruction) {
+      const tagname = tagnameByConstructor.get(this.constructor);
+      const fakeClass = nativeGet.call(window.customElements, tagname);
+
+      // Make sure that the fake constructor doesn't call back to this constructor
+      userConstruction = true;
+      const instance = new (fakeClass)();
+      return instance;
+    }
+    // Else do nothing. This will be reached by ES5-style classes doing
+    // HTMLElement.call() during initialization
+    browserConstruction = false;
+  };
+
+  HTMLElement.prototype = Object.create(NativeHTMLElement.prototype);
+  HTMLElement.prototype.constructor = HTMLElement;
+  HTMLElement.nativeClass = NativeHTMLElement;
+
+  window.customElements.define = (tagname, elementClass) => {
+
+    const StandInElement = class extends NativeHTMLElement {
+      constructor() {
+        // Call the native HTMLElement constructor, this gives us the
+        // under-construction instance as `this`:
+        super();
+
+        // The prototype will be wrong up because the browser used our fake
+        // class, so fix it:
+        Object.setPrototypeOf(this, elementClass.prototype);
+
+        if (!userConstruction) {
+          // Make sure that user-defined constructor bottom's out to a do-nothing
+          // HTMLElement() call
+          browserConstruction = true;
+          // Call the user-defined constructor on our instance:
+          elementClass.call(this);
+        }
+        userConstruction = false;
+      }
+    }
+    if ('observedAttributes' in elementClass) {
+      StandInElement.observedAttributes = elementClass.observedAttributes;
+    }
+    const elementProto = elementClass.prototype;
+    if ('connectedCallback' in elementProto) {
+      StandInElement.prototype.connectedCallback = function() {
```

Comment:
Is it important that this is a new function, rather than just copying connectedCallback off the old proto? Same goes for all the callbacks below.